### PR TITLE
add Parent field to opslevel_service resource

### PR DESCRIPTION
**Pull request by davidbloss**
## Issues

[Add Service's Parent field to config](https://github.com/OpsLevel/team-platform/issues/423)

## Changelog

Add parent (system) field to `opslevel_service` resource

- [X] List your changes here
- [X] Make a `changie` entry

### Given this Terraform config file
```tf
# main.tf
data "opslevel_system" "foo" {
  identifier = "sample_cart"
}

resource "opslevel_service" "foo" {
  name = "foo bar bar baz"

  description = "foo service"
  framework   = "rails"
  language    = "ruby"
  parent      = data.opslevel_system.foo.id

  lifecycle_alias = "beta"
  tier_alias      = "tier_3"
  owner           = "team_platform_3_a"

  api_document_path             = "/swagger.json"
  preferred_api_document_source = "PULL" //or "PUSH"

  aliases = ["foo", "bar", "baz"]
  tags    = ["foo:bar"]
}
```

### Create the `opslevel_service` resource with `parent` field, `terraform apply`
```tf
data.opslevel_system.foo: Reading...
data.opslevel_system.foo: Read complete after 0s [id=Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzEwOTk2NjY]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # opslevel_service.foo will be created
  + resource "opslevel_service" "foo" {
      + aliases                       = [
          + "bar",
          + "baz",
          + "foo",
        ]
      + api_document_path             = "/swagger.json"
      + description                   = "foo service"
      + framework                     = "rails"
      + id                            = (known after apply)
      + language                      = "ruby"
      + lifecycle_alias               = "beta"
      + name                          = "foo bar bar baz"
      + owner                         = "team_platform_3_a"
      + parent                        = "Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzEwOTk2NjY"
      + preferred_api_document_source = "PULL"
      + tags                          = [
          + "foo:bar",
        ]
      + tier_alias                    = "tier_3"
    }

Plan: 1 to add, 0 to change, 0 to destroy.
opslevel_service.foo: Creating...
opslevel_service.foo: Creation complete after 3s [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjc5NTE]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

### Update the Terraform config file, `parent` set to `null`
```tf
data "opslevel_system" "foo" {
  identifier = "sample_cart"
}

resource "opslevel_service" "foo" {
  name = "foo bar bar baz"

  description = "foo service"
  framework   = "rails"
  language    = "ruby"
  parent      = null

  lifecycle_alias = "beta"
  tier_alias      = "tier_3"
  owner           = "team_platform_3_a"

  api_document_path             = "/swagger.json"
  preferred_api_document_source = "PULL" //or "PUSH"

  aliases = ["foo", "bar", "baz"]
  tags    = ["foo:bar"]
}
```

### Update the `opslevel_service` resource (remove `parent`), `terraform apply`
```tf
data.opslevel_system.foo: Reading...
opslevel_service.foo: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjc5NTE]
data.opslevel_system.foo: Read complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzEwOTk2NjY]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # opslevel_service.foo will be updated in-place
  ~ resource "opslevel_service" "foo" {
        id                            = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjc5NTE"
        name                          = "foo bar bar baz"
      - parent                        = "Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzEwOTk2NjY" -> null
        tags                          = [
            "foo:bar",
        ]
        # (9 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
opslevel_service.foo: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjc5NTE]
opslevel_service.foo: Modifications complete after 2s [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjc5NTE]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

### Update the Terraform config file, `parent` set again
```tf
data "opslevel_system" "foo" {
  identifier = "sample_cart"
}

resource "opslevel_service" "foo" {
  name = "foo bar bar baz"

  description = "foo service"
  framework   = "rails"
  language    = "ruby"
  parent      = data.opslevel_system.foo.id

  lifecycle_alias = "beta"
  tier_alias      = "tier_3"
  owner           = "team_platform_3_a"

  api_document_path             = "/swagger.json"
  preferred_api_document_source = "PULL" //or "PUSH"

  aliases = ["foo", "bar", "baz"]
  tags    = ["foo:bar"]
}
```

### Update the `opslevel_service` resource (add `parent` back), `terraform apply`
```tf
data.opslevel_system.foo: Reading...
data.opslevel_system.foo: Read complete after 0s [id=Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzEwOTk2NjY]
opslevel_service.foo: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjc5NTE]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # opslevel_service.foo will be updated in-place
  ~ resource "opslevel_service" "foo" {
        id                            = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjc5NTE"
        name                          = "foo bar bar baz"
      + parent                        = "Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzEwOTk2NjY"
        tags                          = [
            "foo:bar",
        ]
        # (9 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
opslevel_service.foo: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjc5NTE]
opslevel_service.foo: Modifications complete after 2s [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjc5NTE]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

### Destroy the `opslevel_service` resource, `terraform destroy` ✅

### CODE REVIEW

File: opslevel/resource_opslevel_service.go [hunk 1]
```diff
@@ -161,6 +163,11 @@ func (r *ServiceResource) Schema(ctx context.Context, req resource.SchemaRequest
 				Description: "The team that owns the service. ID or Alias may be used.",
 				Optional:    true,
 			},
+			"parent": schema.StringAttribute{
+				Description: "The id of the parent system of this service",
+				Optional:    true,
+				Validators:  []validator.String{IdStringValidator()},
```
Comment (taimoorgit) [inline opslevel/resource_opslevel_service.go:169]: @davidbloss should this validator be here for `parent`? It looks like aliases can be used here just like in `owner`.

**Review by taimoorgit — CHANGES_REQUESTED** [review]
Just 1 blocker

File: opslevel/resource_opslevel_service.go (hunk 1 above)
Comment (davidbloss) [reply opslevel/resource_opslevel_service.go:169]: Great catch 👏  - yes aliases should work too. Updating incoming

**Review by taimoorgit — APPROVED** [review]
@davidbloss I think it is worth adding this into the example for the terraform docs where we have a `opslevel_system` resource along with a `opslevel_service` resource that can be joined together.

File: opslevel/resource_opslevel_service.go [hunk 2]
```diff
@@ -327,12 +338,16 @@ func (r *ServiceResource) Update(ctx context.Context, req resource.UpdateRequest
 		LifecycleAlias: NullableStringConfigValue(planModel.LifecycleAlias),
 		Name:           opslevel.NewNullableFrom(planModel.Name.ValueString()),
 		OwnerInput:     opslevel.NewIdentifier(),
+		Parent:         opslevel.NewIdentifier(),
```
Comment (rocktavious) [inline opslevel/resource_opslevel_service.go:341]: shouldn't this only be set IF the end user has configured it in their plan?

**Review by rocktavious — CHANGES_REQUESTED** [review]
1 edgecase i didn't see in tophatting and yes agree with @taimoorgit this needs to be in the docs example.

File: opslevel/resource_opslevel_service.go (hunk 2 above)
Comment (davidbloss) [reply opslevel/resource_opslevel_service.go:341]: The API won't accept an empty string here so I added a validator to this field. That leaves the following:
- `parent = <value>` the value is set
- `parent = null` no parent is set on create, or the parent is unset on update
- `// parent = <value` same as `null`

Comment (rocktavious) [reply opslevel/resource_opslevel_service.go:347]: Did you test my edgecase?  ```where its always unsetting the value if its set via the UI but not actually in the Terraform config```

**Pull request merged**